File: toris_collection/bird_profile.py
"""bird_profile.py - 図鑑の「プロフィール」行を組み立てる純粋関数。

図鑑に「好きなもの / 好きな場所 / こわいもの」を出すための層。Streamlit に依存しない
(= テストできる)。文言の組み立ては行わず、**構造化データ**を返す。表示側(app.py)が
t() と表示名ヘルパーで描画する — engine が日本語文を組み立てて EN 画面に漏れた
過去の監査指摘(2026-08-02)を繰り返さないため。

原則との関係:
  - 原則4(生態に誠実): 好きなもの = 実際に食べるもの(eats_plants/eats_insects)、
    好きな場所 = 実際の biome_pref、こわいもの = GloBI の実際の捕食記録。
    作り話の「かわいい豆知識」は入れない。
  - 原則2(罰しない): ここは読み物であって、ゲーム進行には一切影響しない。
"""
from __future__ import annotations

import i18n
import predators


def likes(bird: dict, plants: dict, insects: dict) -> list[dict]:
    """「好きなもの」= 実際に食べる植物・昆虫。

    Returns: [{"kind": "plant"|"insect", "id": str, "entity": dict}, ...]
             図鑑に載っていない ID は静かに除外する(データ更新への安全弁)。
    """
    out: list[dict] = []
    for pid in bird.get("eats_plants") or []:
        ent = plants.get(pid)
        if ent:
            out.append({"kind": "plant", "id": pid, "entity": ent})
    for iid in bird.get("eats_insects") or []:
        ent = insects.get(iid)
        if ent:
            out.append({"kind": "insect", "id": iid, "entity": ent})
    return out


def home(bird: dict, biomes: dict) -> list[dict]:
    """「好きな場所」= 実際に好むバイオーム。

    Returns: [{"id": str, "entity": dict}, ...]
    """
    out: list[dict] = []
    for bid in bird.get("biome_pref") or []:
        ent = biomes.get(bid)
        if ent:
            out.append({"id": bid, "entity": ent})
    return out


def fears(bird_id: str) -> dict:
    """「こわいもの」= GloBI の実際の捕食記録(分類カテゴリ)。

    Returns: {"categories": [key, ...], "genus_level": bool}
             データが無ければ categories は空(=図鑑ではこの行を出さない)。
    """
    return {
        "categories": predators.categories(bird_id),
        "genus_level": predators.is_genus_level(bird_id),
    }


def build(bird_id: str, bird: dict, plants: dict, insects: dict,
          biomes: dict) -> dict:
    """図鑑プロフィールの全項目をまとめて返す。

    Returns: {"likes": [...], "home": [...], "fears": {...}}
             各項目は空になりうる(空の行は表示側で出さない=静かなトーン)。
    """
    return {
        "likes": likes(bird, plants, insects),
        "home": home(bird, biomes),
        "fears": fears(bird_id),
    }


def _biome_disp(biome: dict) -> str:
    """バイオームの表示名。英語表示では name_en(無ければ日本語 name)。
    app.py の _biome_display_name と同じ規則(バイオームだけ english ではなく name_en)。"""
    if i18n.get_lang() == "en" and biome.get("name_en"):
        return biome["name_en"]
    return biome.get("name", "")


def rows(bird_id: str, bird: dict, plants: dict, insects: dict,
         biomes: dict) -> list[tuple[str, str, str]]:
    """図鑑プロフィールを表示用の行にする。

    Returns: [(emoji, ラベル, 値), ...] すべて現在の表示言語。
             データの無い項目は行ごと落とす(空欄を並べない=静かなトーン)。

    文言の組み立てをここに集約するのは、Streamlit 抜きでテストできるようにするため
    (EN 画面への日本語漏れを機械的に検出できる状態を保つ)。
    """
    prof = build(bird_id, bird, plants, insects, biomes)
    sep = "、" if i18n.get_lang() == "ja" else ", "
    out: list[tuple[str, str, str]] = []

    if prof["likes"]:
        out.append(("🍽", i18n.t("好きなもの"),
                    sep.join(i18n.disp(x["entity"]) for x in prof["likes"])))

    if prof["home"]:
        out.append(("🏞", i18n.t("好きな場所"),
                    sep.join(_biome_disp(x["entity"]) for x in prof["home"])))

    if prof["fears"]["categories"]:
        value = sep.join(
            i18n.t(predators.CATEGORY_LABELS[k])
            for k in prof["fears"]["categories"]
        )
        if prof["fears"]["genus_level"]:
            # 同属の近縁種の記録から採ったものは、黙って種の事実にしない(原則4)。
            # 区切りの空白も言語に合わせる(全角スペースは EN 画面に混ぜない)。
            gap = "　" if i18n.get_lang() == "ja" else " "
            value += gap + i18n.t("(近縁のなかまの記録から)")
        out.append(("😨", i18n.t("こわいもの"), value))

    return out

File: toris_core/lib/src/bird_profile.dart
/// 図鑑の「好きなもの / 好きな場所 / こわいもの」。
/// `bird_profile.py` と `predators.py` の移植。
///
/// 交渉不能の原則4「生態に誠実」:
///   好きなもの = 実際に食べるもの(eats_plants / eats_insects)
///   好きな場所 = 実際の biome_pref
///   こわいもの = **GloBI の実際の捕食記録**
/// 作り話の「かわいい豆知識」は入れない。
///
/// ## ⚠️ 分類の表を自分で作らないこと(2026-08-16 の監査で見つけた)
/// 私は Dart 側に7つの分類(raptor / owl / snake / **mammal** / cat /
/// **corvid** / **other**)を手で書いていた。実際の `predators.py` は**12分類**で、
/// mammal・corvid・other は**存在しない**。データが使っている
/// crow / falcon / fox / rodent / weasel が表に無く、図鑑には生のキーが
/// そのまま出ていた。表は Python から引き写すこと。
library;

/// 分類キー → 表示名。`predators.CATEGORY_LABELS` の**出荷済みの英語**
/// (`i18n.py` の TRANSLATIONS を引き写した)。
///
/// **順序も Python と同じ**にしてある(表示順ではなく、表の同一性の確認用)。
const Map<String, String> kPredatorLabels = {
  'raptor': 'Hawks',
  'owl': 'Owls',
  'falcon': 'Falcons',
  'snake': 'Snakes',
  'crow': 'Crows and jays',
  'shrike': 'Shrikes',
  'cat': 'Cats',
  'raccoon': 'Raccoons',
  'weasel': 'Weasels',
  'fox': 'Foxes',
  'squirrel': 'Squirrels',
  'rodent': 'Mice and rats',
};

/// その鳥の天敵カテゴリのキー。無ければ空。
///
/// **表に無いキーは静かに落とす**(`predators.categories` と同じ)。
/// データが増えても、知らない分類を画面に出さないための安全弁。
List<String> predatorCategories(
    String birdId, Map<String, dynamic> predatorsData) {
  final rec = predatorsData[birdId] as Map?;
  if (rec == null) return const [];
  return [
    for (final c in ((rec['categories'] as List?) ?? const []))
      if (kPredatorLabels.containsKey('$c')) '$c'
  ];
}

/// 天敵が「同属の近縁種の記録」由来か。
bool predatorIsGenusLevel(String birdId, Map<String, dynamic> predatorsData) {
  final rec = predatorsData[birdId] as Map?;
  return rec != null && rec['level'] == 'genus';
}

/// その鳥の天敵カテゴリを、表示名にして返す。
List<String> predatorLabels(
        String birdId, Map<String, dynamic> predatorsData) =>
    [
      for (final k in predatorCategories(birdId, predatorsData))
        kPredatorLabels[k]!
    ];

/// 天敵のデータがあるか。
bool predatorHasData(String birdId, Map<String, dynamic> predatorsData) =>
    predatorCategories(birdId, predatorsData).isNotEmpty;

/// 「好きなもの」の1件。
class ProfileLike {
  /// 'plant' か 'insect'。
  final String kind;
  final String id;
  const ProfileLike(this.kind, this.id);
}

/// 「好きなもの」= 実際に食べる植物・昆虫。
///
/// **図鑑に載っていない ID は静かに除外する**(データ更新への安全弁)。
/// 並びは Python と同じ「植物 → 昆虫」で、それぞれ元の並び順のまま。
List<ProfileLike> profileLikes(Map<String, dynamic> bird,
    Map<String, dynamic> plants, Map<String, dynamic> insects) {
  final out = <ProfileLike>[];
  for (final p in ((bird['eats_plants'] as List?) ?? const [])) {
    if (plants['$p'] != null) out.add(ProfileLike('plant', '$p'));
  }
  for (final i in ((bird['eats_insects'] as List?) ?? const [])) {
    if (insects['$i'] != null) out.add(ProfileLike('insect', '$i'));
  }
  return out;
}

/// 「好きな場所」= 実際に好む土地。載っていない ID は除外する。
List<String> profileHome(
    Map<String, dynamic> bird, Map<String, dynamic> biomes) {
  final out = <String>[];
  for (final b in ((bird['biome_pref'] as List?) ?? const [])) {
    if (biomes['$b'] != null) out.add('$b');
  }
  return out;
}

/// 「こわいもの」。
class ProfileFears {
  final List<String> categories;
  final bool genusLevel;
  const ProfileFears(this.categories, this.genusLevel);
}

/// 図鑑プロフィールの全項目。`bird_profile.build` と同じ。
///
/// 各項目は空になりうる。**空の行は表示側で出さない**(静かなトーン)。
class BirdProfile {
  final List<ProfileLike> likes;
  final List<String> home;
  final ProfileFears fears;
  const BirdProfile(this.likes, this.home, this.fears);
}

BirdProfile buildBirdProfile({
  required String birdId,
  required Map<String, dynamic> bird,
  required Map<String, dynamic> plants,
  required Map<String, dynamic> insects,
  required Map<String, dynamic> biomes,
  required Map<String, dynamic> predatorsData,
}) =>
    BirdProfile(
      profileLikes(bird, plants, insects),
      profileHome(bird, biomes),
      ProfileFears(predatorCategories(birdId, predatorsData),
          predatorIsGenusLevel(birdId, predatorsData)),
    );

/// 「こわいもの」だけを取り出す。`bird_profile.fears` と同じ。
ProfileFears profileFears(String birdId, Map<String, dynamic> predatorsData) =>
    ProfileFears(predatorCategories(birdId, predatorsData),
        predatorIsGenusLevel(birdId, predatorsData));
